### ISSUE TITLE
Error in print(history)

### ISSUE THREAD

**Issue by HristoBuyukliev** [rstudio/keras#203]
model <- keras_model_sequential()
    model %>%
        layer_batch_normalization(input_shape = c(dim(x_train)[2])) %>%
        layer_dense(units = 20, activation = 'elu', kernel_initializer='glorot_normal') %>%
        layer_dropout(rate = 0.3) %>%
        layer_batch_normalization() %>%
        layer_dense(units = 3, activation = 'linear', kernel_initializer='glorot_normal')
        model %>% compile(
        loss = 'mse',
        optimizer = optimizer_nadam()
    )
    early_stopping <- callback_early_stopping(monitor = 'val_loss', patience = 10)
    save_model_filepath <- '../models/keras/MLP_model.hdf5'
    model_checkpoint <- callback_model_checkpoint(filepath = save_model_filepath, save_best_only = T)
    history_mlp <- model %>% fit(
        x_train, 
        y_train,
        epochs = 100, 
        batch_size = 128, 
        validation_split=0.2, 
        callbacks = c(early_stopping, model_checkpoint),
        verbose = 0
    )

    plot(history_mlp)
    print(history_mlp)

Gives me an error 

      Error in metric[[epochs]] : subscript out of bounds 

Additionally, if I try to debug in browser mode, it behaves very weirdly:
If I do "as.data.frame(history_mlp)" it does not give me an error message or anything, just another prompt for command. However, If press enter, the browser goes a level deeper, like this:

    Browse[1]> history_mlp
    Error in metric[[epochs]] : subscript out of bounds
    Browse[1]> as.data.frame(history_mlp)
    Browse[1]> 
    Browse[2]>

**Comment by HristoBuyukliev** [comment]
note: the output can be shown, and error does not appear, if I write "sink(type="output")".

**Comment by jjallaire** [comment]
Could you provide a complete runnable example so it's easier for us to reproduce the problem?

**Comment by HristoBuyukliev** [comment]
That's it. It only uses x_train and y_train, which are sparse model matrices, generated like so:

    spm <- sparse.model.matrix(~.-1,data=train) %>% log1p
    target_cols = c('x_dim', 'y_dim', 'z_dim')
    x_train = spm[, !(colnames(spm) %in% target_cols)]
    y_train = spm[, colnames(spm) %in% target_cols]

I could give you the full code, but there is really nothing more to it, just some data cleaning and stuff.

**Comment by jjallaire** [comment]
I can't emphasize enough how important it is to provide complete reproducible examples when reporting bugs. Without that the developer has to guess what the rest of your code contains. If it's not readily apparent (e.g. what is the value of `train` above) then you are at best wasting the developers time trying to guess the rest of the code, and at worst simply ensuring that the bug is never fixed because developers don't want to spend indeterminate amounts of time reverse engineering your example.

"really nothing more to it" isn't the rest of the code -- provide all of the code in minimal reproducible  form and most bugs are quite easy to fix! Here you can read up more on this practice: https://stackoverflow.com/questions/5963269/how-to-make-a-great-r-reproducible-example-aka-mcve-minimal-complete-and-ver

**Comment by jjallaire** [comment]
I created my own reproducible example which managed to reproduce the history printing bug but not the `as.data.frame()` behavior you mentioned.

Here is my code:

```r
library(keras)

batch_size <- 128
num_classes <- 10
epochs <- 30

c(c(x_train, y_train), c(x_test, y_test)) %<-% dataset_mnist()

x_train <- array_reshape(x_train, c(nrow(x_train), 784))
x_test <- array_reshape(x_test, c(nrow(x_test), 784))

x_train <- x_train / 255
x_test <- x_test / 255

y_train <- to_categorical(y_train, num_classes)
y_test <- to_categorical(y_test, num_classes)

model <- keras_model_sequential()
model %>% 
  layer_dense(units = 256, activation = 'relu', input_shape = c(784)) %>% 
  layer_dropout(rate = 0.4) %>% 
  layer_dense(units = 128, activation = 'relu') %>%
  layer_dropout(rate = 0.3) %>%
  layer_dense(units = 10, activation = 'softmax')

model %>% compile(
  loss = 'categorical_crossentropy',
  optimizer = optimizer_rmsprop(),
  metrics = c('accuracy')
)

early_stopping <- callback_early_stopping(monitor = 'val_acc')

history <- model %>% fit(
  x_train, y_train,
  batch_size = batch_size,
  epochs = epochs,
  verbose = 1,
  validation_split = 0.2,
  callbacks = c(early_stopping)
)

print(history)
  
as.data.frame(history)
```

This as I said reproduced the history printing problem but not the `as.data.frame` problem. I have just checked in a fix for this here: https://github.com/rstudio/keras/commit/e6d9062b424324859b1df4fcd725b7c210b8276c.

If you can provide a minimal reproducible example for the `as.data.frame` problem I'm sure I can make quick work of that one as well! :-)
